sympy/ntheory/ecm.py

```diff
@@ -0,0 +1,245 @@
+#----------------------------------------------------------------------------#
+#                                                                            #
+#                   Lenstra's Elliptic Curve Factorization                   #
+#                                                                            #
+#----------------------------------------------------------------------------#
+
+
+class Point:
+    """For denoting the point on the curve.
+    For ECM we don't need the y-coordinate. All
+    arithmetic operations use only the x and z coordinates.
+    """
+
+    def __init__(self, x, z):
+        self.x = x
+        self.z = z
+
+def _pt_add(P, Q, diff, n):
+    """
+    Add two points P and Q where diff = P - Q.
+
+    Parameters:
+    ===========
+
+    P : point on the curve in Montgomery form
+    Q : point on the curve in Montgomery form
+    diff = P - Q
+    n = modulus
+
+    References
+    ==========
+
+    .. [1]  http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
+        Algorithm - 3
+    """
+
+    u = (P.x - P.z)*(Q.x + Q.z)
+    v = (P.x + P.z)*(Q.x - Q.z)
+    add, subt = u + v, u - v
+    x_cor = diff.z*add*add % n
+    z_cor = diff.x*subt*subt % n
+    return Point(x_cor, z_cor)
+
+def _pt_double(P, n, a24):
+    """
+    Doubles a point in an elliptic curve in Montgomery form.
+
+    Parameter:
+    ==========
+
+    P : Point on the curve
+    a24 : Parameter for doubling
+    n : modulus
+
+    References
+    ==========
+
+    .. [1]  http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
+        Algorithm - 3
+    """
+
+    u, v = P.x + P.z, P.x - P.z
+    u, v = u*u, v*v
+    diff = u - v
+    x_cor = u*v % n
+    z_cor = diff*(v + a24*diff) % n
+    return Point(x_cor, z_cor)
+
+def _mont_ladder(k, P, n, a24):
```

Comment:
And here `P` should be the first parameter: `def mont_ladder(P, k, n, a24):`

Reply:
we may also define `__mul__` method then we would not have to use `double` and `mont_ladder` outside the class.